express/manifest.py

```diff
@@ -1,112 +1,113 @@
-        type (Union[DataType, str]): the type of data, which determines how it's interpreted
-        extensions (List[str]): under what file extensions the data is stored.
-        n_files (int): how many files make up the data
-        n_items (int): how many items (e.g. distinct images, captions, etc.) are covered by the
-         data.
+        specification: The part of the manifest json representing the subset
+        base_path: The base path which the subset location is defined relative to
     """
 
-    location: str
-    type: DataType
-    extensions: List[str]
-    n_files: int = field(default_factory=int)
-    n_items: int = field(default_factory=int)
+    def __init__(self, specification: dict, *, base_path: str) -> None:
+        self._specification = specification
+        self._base_path = base_path
 
+    @property
+    def location(self) -> str:
+        """The resolved location of the subset"""
+        return self._base_path.rstrip("/") + self._specification["location"]
 
-@dataclass_json
-@dataclass
-class Metadata:
-    """
-    The metadata associated with the manifest
-    Args:
-        artifact_bucket (str): remote location of all newly created job artifacts.
-        run_id (str): the kfp run id associated with the manifest (kfp.dsl.EXECUTION_ID_PLACEHOLDER)
-        component_id (str): if this metadata is passed as an input to a component, this is the id of
-         the current component. If this metadata is stored as part of a manifest, this is the id of
-        the component that generated the manifest.
-        component_name (str): name of the current or originating component (see component_id for
-         more details).
-        branch (str): the git branch associated with that manifest
-        commit_hash (str): the commit hash associated with that manifest
-        creation_date (str): the creation date of the manifest
-        num_items (int): total number of rows in the index
-    """
+    @property
+    def fields(self) -> t.Dict[str, Field]:
+        return {
+            name: Field(name=name, type=field["type"])
+            for name, field in self._specification["fields"].items()
+        }
 
-    # TODO: get rid of defaults
-    artifact_bucket: str = field(default_factory=str)
-    run_id: str = field(default_factory=str)
-    component_id: str = field(default_factory=str)
-    component_name: str = field(default_factory=str)
-    branch: str = field(default_factory=str)
-    commit_hash: str = field(default_factory=str)
-    creation_date: str = field(default_factory=str)
-    num_items: int = field(default_factory=int)
 
+class Index(Subset):
+    """Special case of a subset for the index, which has fixed fields"""
 
-@dataclass_json
-@dataclass
-class DataManifest:
+    @property
+    def fields(self) -> t.Dict[str, Field]:
+        return {
+            "id": Field(name="id", type="str"),
+            "source": Field(name="source", type="str"),
```

Comment:
If I understand correctly, "source" is the same thing as "namespace"?

Reply:
Yes, both terms have been used interchangeably.